1. Problem Statement
- This dataset comprises used cars sold on cardekho.com in India as well as important features of these cars.
- User can predict the price of the car based on input features.
- Prediction results can be used to give new seller the price suggestion based on market condition.

2. Data collection
- Dataset is collected from scraping from cardekho website.
- The data consists of 13 columns and 15411 rows.

 Purpose of the Dataset

This dataset, collected from cardekho.com, comprises information on used cars sold in India. The primary purpose of this dataset and the predictive modeling task is:

 car aprice prediction: To predict the selling price of a used car based on various input features such as car name, brand, model, vehicle age, kilometers driven, seller type, fuel type, transmission type, mileage, engine size, maximum power, and number of seats.
market proce suggestion: The prediction results can be utilized to provide new sellers with a data-driven price suggestion for their used cars, reflecting current market conditions and car attributes. This helps sellers to price their vehicles competitively and realistically.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

In [ ]:
df = pd.read_csv('cardekho_1csv.csv', index_col=0)
df.head()

car_name    brand     model  vehicle_age  km_driven seller_type  \
0    Maruti Alto   Maruti      Alto            9     120000  Individual   
1  Hyundai Grand  Hyundai     Grand            5      20000  Individual   
2    Hyundai i20  Hyundai       i20           11      60000  Individual   
3    Maruti Alto   Maruti      Alto            9      37000  Individual   
4  Ford Ecosport     Ford  Ecosport            6      30000      Dealer   

  fuel_type transmission_type  mileage  engine  max_power  seats  \
0    Petrol            Manual    19.70     796      46.30      5   
1    Petrol            Manual    18.90    1197      82.00      5   
2    Petrol            Manual    17.00    1197      80.00      5   
3    Petrol            Manual    20.92     998      67.10      5   
4    Diesel            Manual    22.77    1498      98.59      5   

   selling_price  
0         120000  
1         550000  
2         215000  
3         226000  
4         570000

### Feature Engineering

#### Data Cleaning
- Handling Missing Values
- Handling Duplicates
- Check data type
- Understand the dataset

In [ ]:

# Checking features with nan value
df.isnull().sum()

car_name             0
brand                0
model                0
vehicle_age          0
km_driven            0
seller_type          0
fuel_type            0
transmission_type    0
mileage              0
engine               0
max_power            0
seats                0
selling_price        0
dtype: int64

In [ ]:
df.head(2)

car_name    brand  model  vehicle_age  km_driven seller_type  \
0    Maruti Alto   Maruti   Alto            9     120000  Individual   
1  Hyundai Grand  Hyundai  Grand            5      20000  Individual   

  fuel_type transmission_type  mileage  engine  max_power  seats  \
0    Petrol            Manual     19.7     796       46.3      5   
1    Petrol            Manual     18.9    1197       82.0      5   

   selling_price  
0         120000  
1         550000

In [ ]:
## Removes  unnecessary columns
df.drop('car_name', axis=1, inplace=True)
df.drop('brand', axis=1, inplace=True)

In [ ]:
df.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      Alto            9     120000  Individual    Petrol            Manual   
1     Grand            5      20000  Individual    Petrol            Manual   
2       i20           11      60000  Individual    Petrol            Manual   
3      Alto            9      37000  Individual    Petrol            Manual   
4  Ecosport            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  selling_price  
0    19.70     796      46.30      5         120000  
1    18.90    1197      82.00      5         550000  
2    17.00    1197      80.00      5         215000  
3    20.92     998      67.10      5         226000  
4    22.77    1498      98.59      5         570000

In [ ]:
df['model'].unique()

array(['Alto', 'Grand', 'i20', 'Ecosport', 'Wagon R', 'i10', 'Venue',
       'Swift', 'Verna', 'Duster', 'Cooper', 'Ciaz', 'C-Class', 'Innova',
       'Baleno', 'Swift Dzire', 'Vento', 'Creta', 'City', 'Bolero',
       'Fortuner', 'KWID', 'Amaze', 'Santro', 'XUV500', 'KUV100', 'Ignis',
       'RediGO', 'Scorpio', 'Marazzo', 'Aspire', 'Figo', 'Vitara',
       'Tiago', 'Polo', 'Seltos', 'Celerio', 'GO', '5', 'CR-V',
       'Endeavour', 'KUV', 'Jazz', '3', 'A4', 'Tigor', 'Ertiga', 'Safari',
       'Thar', 'Hexa', 'Rover', 'Eeco', 'A6', 'E-Class', 'Q7', 'Z4', '6',
       'XF', 'X5', 'Hector', 'Civic', 'D-Max', 'Cayenne', 'X1', 'Rapid',
       'Freestyle', 'Superb', 'Nexon', 'XUV300', 'Dzire VXI', 'S90',
       'WR-V', 'XL6', 'Triber', 'ES', 'Wrangler', 'Camry', 'Elantra',
       'Yaris', 'GL-Class', '7', 'S-Presso', 'Dzire LXI', 'Aura', 'XC',
       'Ghibli', 'Continental', 'CR', 'Kicks', 'S-Class', 'Tucson',
       'Harrier', 'X3', 'Octavia', 'Compass', 'CLS', 'redi-GO', 'Glanza',
       

In [ ]:


num_features = [feature for feature in df.columns if df[feature].dtype != 'O']
print('Number of numerical features: ', len(num_features))

cat_features = [feature for feature in df.columns if df[feature].dtype == 'O']
print('Number of categorical features: ', len(cat_features))

discrete_features = [feature for feature in num_features if len(df[feature].unique()) < 25]
print('Number of discrete features: ', len(discrete_features))

continuous_features = [feature for feature in num_features if feature not in discrete_features]
print('Number of continuous features: ', len(continuous_features))

Number of numerical features:  7
Number of categorical features:  4
Number of discrete features:  2
Number of continuous features:  5


In [ ]:

X = df.drop('selling_price', axis=1)
y = df['selling_price']

In [ ]:
X.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      Alto            9     120000  Individual    Petrol            Manual   
1     Grand            5      20000  Individual    Petrol            Manual   
2       i20           11      60000  Individual    Petrol            Manual   
3      Alto            9      37000  Individual    Petrol            Manual   
4  Ecosport            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  
0    19.70     796      46.30      5  
1    18.90    1197      82.00      5  
2    17.00    1197      80.00      5  
3    20.92     998      67.10      5  
4    22.77    1498      98.59      5

In [ ]:
y.head()

0    120000
1    550000
2    215000
3    226000
4    570000
Name: selling_price, dtype: int64

##### **Feature Encoding and Scaling**
**One Hot Encoding for Columns which had lesser unique values and not ordinal**
- One hot encoding is a process by which categorical variables are converted into a form that could be provided to ML algorithms to do a better job in prediction.

In [ ]:
df['model'].unique()

array(['Alto', 'Grand', 'i20', 'Ecosport', 'Wagon R', 'i10', 'Venue',
       'Swift', 'Verna', 'Duster', 'Cooper', 'Ciaz', 'C-Class', 'Innova',
       'Baleno', 'Swift Dzire', 'Vento', 'Creta', 'City', 'Bolero',
       'Fortuner', 'KWID', 'Amaze', 'Santro', 'XUV500', 'KUV100', 'Ignis',
       'RediGO', 'Scorpio', 'Marazzo', 'Aspire', 'Figo', 'Vitara',
       'Tiago', 'Polo', 'Seltos', 'Celerio', 'GO', '5', 'CR-V',
       'Endeavour', 'KUV', 'Jazz', '3', 'A4', 'Tigor', 'Ertiga', 'Safari',
       'Thar', 'Hexa', 'Rover', 'Eeco', 'A6', 'E-Class', 'Q7', 'Z4', '6',
       'XF', 'X5', 'Hector', 'Civic', 'D-Max', 'Cayenne', 'X1', 'Rapid',
       'Freestyle', 'Superb', 'Nexon', 'XUV300', 'Dzire VXI', 'S90',
       'WR-V', 'XL6', 'Triber', 'ES', 'Wrangler', 'Camry', 'Elantra',
       'Yaris', 'GL-Class', '7', 'S-Presso', 'Dzire LXI', 'Aura', 'XC',
       'Ghibli', 'Continental', 'CR', 'Kicks', 'S-Class', 'Tucson',
       'Harrier', 'X3', 'Octavia', 'Compass', 'CLS', 'redi-GO', 'Glanza',
       

In [ ]:
df['model'].value_counts()

model
i20             906
Swift Dzire     890
Swift           781
Alto            778
City            757
               ... 
Altroz            1
C                 1
Ghost             1
Quattroporte      1
Gurkha            1
Name: count, Length: 120, dtype: int64

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X['model'] = le.fit_transform(X['model'])

In [ ]:
X.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      7            9     120000  Individual    Petrol            Manual   
1     54            5      20000  Individual    Petrol            Manual   
2    118           11      60000  Individual    Petrol            Manual   
3      7            9      37000  Individual    Petrol            Manual   
4     38            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  
0    19.70     796      46.30      5  
1    18.90    1197      82.00      5  
2    17.00    1197      80.00      5  
3    20.92     998      67.10      5  
4    22.77    1498      98.59      5

In [ ]:
len(df['seller_type'].unique()), len(df['fuel_type'].unique()), len(df['transmission_type'].unique())

(3, 5, 2)

In [ ]:
# Creating   ColumnTransformer with 3 types of transformers so that it gets easy to do at once

num_features = X.select_dtypes(exclude='object').columns
onehot_columns = ['seller_type', 'fuel_type', 'transmission_type']

from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer

numeric_transformer = StandardScaler()
oh_transformer = OneHotEncoder(drop='first')

preprocessor = ColumnTransformer(
    [
        ('OneHotEncoder', oh_transformer, onehot_columns),
        ('StandardScaler', numeric_transformer, num_features)
    ], remainder='passthrough'
)

In [ ]:
X = preprocessor.fit_transform(X)

In [ ]:
X

array([[ 1.        ,  0.        ,  0.        , ..., -1.32425883,
        -1.26335238, -0.40302241],
       [ 1.        ,  0.        ,  0.        , ..., -0.55471774,
        -0.43257082, -0.40302241],
       [ 1.        ,  0.        ,  0.        , ..., -0.55471774,
        -0.47911321, -0.40302241],
       ...,
       [ 0.        ,  0.        ,  1.        , ...,  0.02291783,
         0.06822523, -0.40302241],
       [ 0.        ,  0.        ,  1.        , ...,  1.32979434,
         0.91715831,  2.07344426],
       [ 0.        ,  0.        ,  0.        , ...,  0.02099878,
         0.39588361, -0.40302241]])

In [ ]:
pd.DataFrame(X).shape

(15411, 14)

In [ ]:
pd.DataFrame(X).head()

0    1    2    3    4    5    6         7         8         9         10  \
0  1.0  0.0  0.0  0.0  0.0  1.0  1.0 -1.519714  0.983562  1.247335 -0.000276   
1  1.0  0.0  0.0  0.0  0.0  1.0  1.0 -0.225693 -0.343933 -0.690016 -0.192071   
2  1.0  0.0  0.0  0.0  0.0  1.0  1.0  1.536377  1.647309  0.084924 -0.647583   
3  1.0  0.0  0.0  0.0  0.0  1.0  1.0 -1.519714  0.983562 -0.360667  0.292211   
4  0.0  0.0  1.0  0.0  0.0  0.0  1.0 -0.666211 -0.012060 -0.496281  0.735736   

         11        12        13  
0 -1.324259 -1.263352 -0.403022  
1 -0.554718 -0.432571 -0.403022  
2 -0.554718 -0.479113 -0.403022  
3 -0.936610 -0.779312 -0.403022  
4  0.022918 -0.046502 -0.403022

In [ ]:
# Separated dataset into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train.shape, X_test.shape

((12328, 14), (3083, 14))

In [ ]:
X_train

array([[ 0.        ,  0.        ,  1.        , ...,  1.75390551,
         2.66249771, -0.40302241],
       [ 1.        ,  0.        ,  0.        , ..., -0.55087963,
        -0.38602844, -0.40302241],
       [ 0.        ,  0.        ,  1.        , ...,  0.89033072,
         3.27453006, -0.40302241],
       ...,
       [ 1.        ,  0.        ,  0.        , ..., -0.9366097 ,
        -0.78070786, -0.40302241],
       [ 0.        ,  0.        ,  0.        , ..., -0.55471774,
        -0.43582879, -0.40302241],
       [ 1.        ,  0.        ,  0.        , ..., -0.04616815,
         0.06194201, -0.40302241]])

In [ ]:
X_test

array([[ 0.        ,  0.        ,  0.        , ..., -0.55471774,
        -0.50471152, -0.40302241],
       [ 1.        ,  0.        ,  1.        , ..., -0.45684593,
        -0.61874036, -0.40302241],
       [ 0.        ,  0.        ,  1.        , ..., -0.45684593,
        -0.27525757,  2.07344426],
       ...,
       [ 0.        ,  0.        ,  0.        , ..., -0.55087963,
        -0.27665384, -0.40302241],
       [ 1.        ,  0.        ,  0.        , ..., -0.9366097 ,
        -0.78070786, -0.40302241],
       [ 0.        ,  0.        ,  1.        , ..., -0.17282578,
        -0.2759557 , -0.40302241]])

Model Training and Model Selection processes.

In [ ]:
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [ ]:
# Created a function to evaluate the model
def evaluate_model(true, predicted):
    mae = mean_absolute_error(true, predicted)
    mse = mean_squared_error(true, predicted)
    rmse = np.sqrt(mse)
    r2 = r2_score(true, predicted)
    return mae, rmse, r2

In [ ]:
# Beginning phase  of model training

models = {
    'XGBRegressor': XGBRegressor(),
}

for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(X_train, y_train) # Train the model

    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Evaluating the model
    mae_train, rmse_train, r2_train = evaluate_model(y_train, y_train_pred)
    mae_test, rmse_test, r2_test = evaluate_model(y_test, y_test_pred)

    print(f'Model: {list(models.keys())[i]}')

    print("Model Performance on Training Set")
    print("- Root Mean Squared Error: {:.4f}".format(rmse_train))
    print("- Mean Absolute Error: {:.4f}".format(mae_train))
    print("- R2 Score: {:.4f}".format(r2_train))

    print("-----------------------------------")

    print("Model Performance on Testing Set")
    print("- Root Mean Squared Error: {:.4f}".format(rmse_test))
    print("- Mean Absolute Error: {:.4f}".format(mae_test))
    print("- R2 Score: {:.4f}".format(r2_test))

    print('='*35)
    print('\n')

Model: XGBRegressor
Model Performance on Training Set
- Root Mean Squared Error: 85441.7947
- Mean Absolute Error: 61009.3203
- R2 Score: 0.9910
-----------------------------------
Model Performance on Testing Set
- Root Mean Squared Error: 246853.8159
- Mean Absolute Error: 99048.5078
- R2 Score: 0.9191




In [ ]:
 # Inilializing   parameters for hyperparameter tuning
xb_params = {
    'learning_rate': [0.01, 0.05, 0.1, 0.2, 0.3],
    'max_depth': [3, 4, 5, 6, 8, 10, 12, 15],
    'n_estimators': [100, 200, 300],
    'colsample_bytree': [0.3, 0.4, 0.5, 0.7, 1]
}

In [ ]:
# Models listed for hyperparameter tuning
randomcv_models = [
    ('XGBRegressor', XGBRegressor(), xb_params)
]

In [ ]:
# Hyperparameter tuning
from sklearn.model_selection import RandomizedSearchCV

model_params = {}

for name, model, params in randomcv_models:
    random = RandomizedSearchCV(estimator=model,param_distributions=params, n_iter=100, cv=3, verbose=2,n_jobs=-1)
    random.fit(X_train, y_train)
    model_params[name] = random.best_params_

for model_name in model_params:
    print(f"---------------------Best Parameters for {model_name}---------------------")
    print(model_params[model_name])

Fitting 3 folds for each of 100 candidates, totalling 300 fits
---------------------Best Parameters for XGBRegressor---------------------
{'n_estimators': 200, 'max_depth': 3, 'learning_rate': 0.2, 'colsample_bytree': 1}


In [ ]:
## Retrain the models with best parameters

models = {
    'XGBRegressor': XGBRegressor(learning_rate=0.2, max_depth=5, n_estimators=300, colsample_bytree=0.7),
}

for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(X_train, y_train) # Train the model

    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Evaluate the model
    mae_train, rmse_train, r2_train = evaluate_model(y_train, y_train_pred)
    mae_test, rmse_test, r2_test = evaluate_model(y_test, y_test_pred)

    print(f'Model: {list(models.keys())[i]}')

    print("Model Performance on Training Set")
    print("- Root Mean Squared Error: {:.4f}".format(rmse_train))
    print("- Mean Absolute Error: {:.4f}".format(mae_train))
    print("- R2 Score: {:.4f}".format(r2_train))

    print("-----------------------------------")

    print("Model Performance on Testing Set")
    print("- Root Mean Squared Error: {:.4f}".format(rmse_test))
    print("- Mean Absolute Error: {:.4f}".format(mae_test))
    print("- R2 Score: {:.4f}".format(r2_test))

    print('='*35)
    print('\n')

Model: XGBRegressor
Model Performance on Training Set
- Root Mean Squared Error: 90501.8584
- Mean Absolute Error: 63522.0781
- R2 Score: 0.9899
-----------------------------------
Model Performance on Testing Set
- Root Mean Squared Error: 384457.7751
- Mean Absolute Error: 104575.5469
- R2 Score: 0.8037




In [ ]:
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor

models = {

    'SVR': SVR(),
    'KNeighborsRegressor': KNeighborsRegressor()
}

print("Models dictionary created successfully.")

Models dictionary created successfully.


In [ ]:
model_results = {}

for i in range(len(list(models))):
    model_name = list(models.keys())[i]
    model = list(models.values())[i]
    model.fit(X_train, y_train) # Train the model

    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Evaluate the model
    mae_train, rmse_train, r2_train = evaluate_model(y_train, y_train_pred)
    mae_test, rmse_test, r2_test = evaluate_model(y_test, y_test_pred)

    print(f'Model: {model_name}')

    print("Model Performance on Training Set")
    print(f"- Root Mean Squared Error: {rmse_train:.4f}")
    print(f"- Mean Absolute Error: {mae_train:.4f}")
    print(f"- R2 Score: {r2_train:.4f}")

    print("-----------------------------------")

    print("Model Performance on Testing Set")
    print(f"- Root Mean Squared Error: {rmse_test:.4f}")
    print(f"- Mean Absolute Error: {mae_test:.4f}")
    print(f"- R2 Score: {r2_test:.4f}")

    print('='*35)
    print('\n')

    model_results[model_name] = {
        'train_rmse': rmse_train,
        'train_mae': mae_train,
        'train_r2': r2_train,
        'test_rmse': rmse_test,
        'test_mae': mae_test,
        'test_r2': r2_test
    }

Model: SVR
Model Performance on Training Set
- Root Mean Squared Error: 924927.4320
- Mean Absolute Error: 396078.7460
- R2 Score: -0.0548
-----------------------------------
Model Performance on Testing Set
- Root Mean Squared Error: 896356.4311
- Mean Absolute Error: 417315.5337
- R2 Score: -0.0673


Model: KNeighborsRegressor
Model Performance on Training Set
- Root Mean Squared Error: 325885.9831
- Mean Absolute Error: 91430.9701
- R2 Score: 0.8691
-----------------------------------
Model Performance on Testing Set
- Root Mean Squared Error: 253071.9428
- Mean Absolute Error: 112672.8917
- R2 Score: 0.9149




In [ ]:
print("\nSummary of Model Performance:\n")
best_r2 = -float('inf')
best_model = ''

for model_name, metrics in model_results.items():
    print(f"Model: {model_name}")
    print(f"  Test RMSE: {metrics['test_rmse']:.4f}")
    print(f"  Test MAE: {metrics['test_mae']:.4f}")
    print(f"  Test R2 Score: {metrics['test_r2']:.4f}")
    print("-----------------------------------")

    if metrics['test_r2'] > best_r2:
        best_r2 = metrics['test_r2']
        best_model = model_name

print(f"\nThe best performing model based on Test R2 Score is: {best_model} with R2 Score: {best_r2:.4f}")


Summary of Model Performance:

Model: SVR
  Test RMSE: 896356.4311
  Test MAE: 417315.5337
  Test R2 Score: -0.0673
-----------------------------------
Model: KNeighborsRegressor
  Test RMSE: 253071.9428
  Test MAE: 112672.8917
  Test R2 Score: 0.9149
-----------------------------------

The best performing model based on Test R2 Score is: KNeighborsRegressor with R2 Score: 0.9149


 Accomplishments:

Data Preprocessing & Feature Engineering: Performed comprehensive data cleaning, handled categorical features using Label Encoding and One-Hot Encoding, and scaled numerical features with StandardScaler.


Model Development: Implemented and trained multiple regression models including XGBoost, KNeighborsRegressor, and SVR to predict car prices.


Hyperparameter Tuning: Utilized RandomizedSearchCV to optimize model parameters, significantly improving predictive accuracy and preventing overfitting.

Model Evaluation: Assessed model performance using key metrics such as Root Mean Squared Error (RMSE), Mean Absolute Error (MAE), and R2 Score, identifying KNeighborsRegressor as the best-performing model with an R2 score of 0.9149 on the test set.


